File: apps/shared/selected-session-proxy.mjs
const prefix = '/api/selected-node/'
const selectedCookie = '__Host-hl-selected'
const antiforgeryHeader = 'x-harborline-antiforgery'
const maximumBodyBytes = 10 * 1024 * 1024
const responseHeaders = ['content-type', antiforgeryHeader, 'x-harborline-audit-id', 'x-harborline-audit-correlation', 'x-correlation-id']

/** A separate, cookie-only transport. Never reuse the bootstrap proxy's headers or cookie state. */
export function createSelectedSessionProxy(nodeOrigin) {
  const target = new URL(nodeOrigin)
  if (!['http:', 'https:'].includes(target.protocol) || target.username || target.password
    || target.pathname !== '/' || target.search || target.hash) throw new Error('A fixed local-node origin is required.')

  return async (request, response, next) => {
    if (!request.url?.startsWith(prefix)) return next()
    const refuse = (status, code) => response.writeHead(status, {
      'Content-Type': 'application/json', 'Cache-Control': 'no-store',
    }).end(JSON.stringify({ error: code }))
    const path = request.url.slice(prefix.length)
    const pathname = path.split('?')[0]
    if (!/^(local-node|session)\//.test(pathname) || /[\\#\x00-\x20]|%2e|%2f|%5c/i.test(pathname)
      || pathname.includes('//') || pathname.split('/').some(part => part === '.' || part === '..'))
      return refuse(400, 'selected_session_path_invalid')

    const method = request.method ?? 'GET'
    if (!['GET', 'POST', 'PUT', 'PATCH', 'DELETE'].includes(method)) return refuse(405, 'selected_session_method_invalid')
    const read = method === 'GET'
    const appOrigin = `${request.socket.encrypted ? 'https' : 'http'}://${request.headers.host}`
    if ((request.headers.origin && request.headers.origin !== appOrigin)
      || (request.headers['sec-fetch-site'] && request.headers['sec-fetch-site'] !== 'same-origin')
      || (!request.headers.origin && (!read || request.headers['sec-fetch-site'] !== 'same-origin')))
      return refuse(403, 'selected_session_origin_invalid')

    const cookies = (request.headers.cookie ?? '').split(';').map(part => part.trim())
      .filter(part => part.startsWith(`${selectedCookie}=`))
    if (cookies.length !== 1 || !/^__Host-hl-selected=[A-Za-z0-9_-]{1,256}$/.test(cookies[0]))
      return refuse(401, 'selected_session_required')
    const token = request.headers[antiforgeryHeader]
    if (!read && (typeof token !== 'string' || !/^[A-Za-z0-9_-]{1,256}$/.test(token)))
      return refuse(403, 'selected_session_antiforgery_required')
    const requestId = request.headers['idempotency-key']
    if (requestId !== undefined && (typeof requestId !== 'string' || !/^[A-Za-z0-9._:-]{1,128}$/.test(requestId)))
      return refuse(400, 'selected_session_headers_invalid')
    const correlationId = request.headers['x-correlation-id']
    if (correlationId !== undefined && (typeof correlationId !== 'string'
      || !/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i.test(correlationId)
      || correlationId === '00000000-0000-0000-0000-000000000000'))
      return refuse(400, 'selected_session_headers_invalid')

    // Positive allowlist: neither inbound Authorization nor server bootstrap credentials can flow.
    const headers = { Cookie: cookies[0], Accept: 'application/json' }
    if (requestId !== undefined) headers['Idempotency-Key'] = requestId
    if (correlationId !== undefined) headers['X-Correlation-ID'] = correlationId
    if (!read) headers[antiforgeryHeader] = token
    if (request.headers['content-type']) headers['Content-Type'] = request.headers['content-type']
    try {
      let body
      if (!read) {
        const chunks = []
        let length = 0
        for await (const chunk of request) {
          length += chunk.length
          if (length > maximumBodyBytes) return refuse(413, 'selected_session_body_too_large')
          chunks.push(chunk)
        }
        body = Buffer.concat(chunks)
      }
      const upstream = await fetch(new URL(`/api/${path}`, target), { method, headers, body, redirect: 'manual' })
      // Redirects must not move a credentialed request outside this configured node.
      if (upstream.status >= 300 && upstream.status < 400) return refuse(502, 'selected_session_redirect_refused')
      const outgoing = { 'Cache-Control': 'no-store' }
      for (const name of responseHeaders) if (upstream.headers.has(name)) outgoing[name] = upstream.headers.get(name)
      const selectedCookies = upstream.headers.getSetCookie().filter(value => value.startsWith(`${selectedCookie}=`))
      if (selectedCookies.length) outgoing['Set-Cookie'] = selectedCookies
      const bytes = new Uint8Array(await upstream.arrayBuffer())
      response.writeHead(upstream.status, outgoing).end(bytes)
    } catch {
      return refuse(502, 'selected_session_upstream_unavailable')
    }
  }
}

File: apps/blazor/Transport/SelectedSessionProxy.cs
using System.Text.RegularExpressions;
using Microsoft.AspNetCore.Http.Features;

namespace Harborline.App.Blazor.ReferenceHost.Transport;

/// <summary>Browser-cookie-only forwarding, isolated from every bootstrap HttpClient.</summary>
public sealed class SelectedSessionProxy : IDisposable
{
    public const string Route = "/api/selected-node/{**path}";
    private const string Prefix = "/api/selected-node/";
    private const string CookieName = "__Host-hl-selected";
    private const string AntiforgeryHeader = "X-Harborline-Antiforgery";
    private const int MaximumBodyBytes = 10 * 1024 * 1024;
    private static readonly Regex Token = new("^[A-Za-z0-9_-]{1,256}$", RegexOptions.CultureInvariant);
    private static readonly Regex RequestId = new("^[A-Za-z0-9._:-]{1,128}$", RegexOptions.CultureInvariant);
    private static readonly Regex PathEscape = new(@"[\\#\x00-\x20]|%2e|%2f|%5c", RegexOptions.IgnoreCase | RegexOptions.CultureInvariant);
    private static readonly string[] ResponseHeaders = [AntiforgeryHeader, "X-Harborline-Audit-Id", "X-Harborline-Audit-Correlation", "X-Correlation-Id"];
    private readonly Uri _origin;
    private readonly HttpMessageInvoker _sender;

    public SelectedSessionProxy(Uri origin) : this(origin, new HttpMessageInvoker(new SocketsHttpHandler
    {
        UseCookies = false,
        AllowAutoRedirect = false,
    })) { }

    public SelectedSessionProxy(Uri origin, HttpMessageInvoker sender)
    {
        ArgumentNullException.ThrowIfNull(origin);
        ArgumentNullException.ThrowIfNull(sender);
        if (!origin.IsAbsoluteUri || (origin.Scheme != "http" && origin.Scheme != "https")
            || origin.UserInfo.Length != 0 || origin.AbsolutePath != "/" || origin.Query.Length != 0 || origin.Fragment.Length != 0)
            throw new ArgumentException("A fixed local-node origin is required.", nameof(origin));
        _origin = origin;
        _sender = sender;
    }

    public async Task ForwardAsync(HttpContext context)
    {
        ArgumentNullException.ThrowIfNull(context);
        var request = context.Request;
        var raw = context.Features.Get<IHttpRequestFeature>()?.RawTarget;
        if (string.IsNullOrEmpty(raw)) raw = request.Path.Value + request.QueryString.Value;
        var path = raw.StartsWith(Prefix, StringComparison.Ordinal) ? raw[Prefix.Length..] : string.Empty;
        var pathname = path.Split('?', 2)[0];
        if ((!pathname.StartsWith("local-node/", StringComparison.Ordinal) && !pathname.StartsWith("session/", StringComparison.Ordinal))
            || PathEscape.IsMatch(pathname) || pathname.Contains("//", StringComparison.Ordinal)
            || pathname.Split('/').Any(part => part is "." or ".."))
        {
            await RefuseAsync(context, 400, "selected_session_path_invalid");
            return;
        }
        if (request.Method is not ("GET" or "POST" or "PUT" or "PATCH" or "DELETE"))
        {
            await RefuseAsync(context, 405, "selected_session_method_invalid");
            return;
        }
        var read = request.Method == "GET";
        var origin = request.Headers.Origin.ToString();
        var fetchSite = request.Headers["Sec-Fetch-Site"].ToString();
        var expectedOrigin = request.Scheme + "://" + request.Host;
        if ((origin.Length != 0 && !string.Equals(origin, expectedOrigin, StringComparison.Ordinal))
            || (fetchSite.Length != 0 && fetchSite != "same-origin")
            || (origin.Length == 0 && (!read || fetchSite != "same-origin")))
        {
            await RefuseAsync(context, 403, "selected_session_origin_invalid");
            return;
        }
        var cookies = request.Headers.Cookie.ToString().Split(';').Select(part => part.Trim())
            .Where(part => part.StartsWith(CookieName + "=", StringComparison.Ordinal)).ToArray();
        if (cookies.Length != 1 || !Token.IsMatch(cookies[0][(CookieName.Length + 1)..]))
        {
            await RefuseAsync(context, 401, "selected_session_required");
            return;
        }
        var token = request.Headers[AntiforgeryHeader].ToString();
        if (!read && !Token.IsMatch(token))
        {
            await RefuseAsync(context, 403, "selected_session_antiforgery_required");
            return;
        }

        var requestIds = request.Headers["Idempotency-Key"];
        if (requestIds.Count > 1 || (requestIds.Count == 1 && !RequestId.IsMatch(requestIds.ToString())))
        {
            await RefuseAsync(context, 400, "selected_session_headers_invalid");
            return;
        }
        using var outgoing = new HttpRequestMessage(new HttpMethod(request.Method), new Uri(_origin, "/api/" + path));
        if (requestIds.Count == 1) outgoing.Headers.Add("Idempotency-Key", requestIds.ToString());
        outgoing.Headers.Add("Cookie", cookies[0]);
        outgoing.Headers.Add("Accept", "application/json");
        if (!read)
        {
            outgoing.Headers.Add(AntiforgeryHeader, token);
            using var body = new MemoryStream();
            var buffer = new byte[16384];
            int bytes;
            while ((bytes = await request.Body.ReadAsync(buffer, context.RequestAborted)) != 0)
            {
                if (body.Length + bytes > MaximumBodyBytes)
                {
                    await RefuseAsync(context, 413, "selected_session_body_too_large");
                    return;
                }
                body.Write(buffer, 0, bytes);
            }
            outgoing.Content = new ByteArrayContent(body.ToArray());
            if (request.ContentType is not null) outgoing.Content.Headers.TryAddWithoutValidation("Content-Type", request.ContentType);
        }
        try
        {
            using var upstream = await _sender.SendAsync(outgoing, context.RequestAborted);
            if ((int)upstream.StatusCode is >= 300 and < 400)
            {
                await RefuseAsync(context, 502, "selected_session_redirect_refused");
                return;
            }
            context.Response.StatusCode = (int)upstream.StatusCode;
            context.Response.Headers.CacheControl = "no-store";
            context.Response.ContentType = upstream.Content.Headers.ContentType?.ToString();
            foreach (var name in ResponseHeaders)
                if (upstream.Headers.TryGetValues(name, out var values)) context.Response.Headers[name] = values.ToArray();
            if (upstream.Headers.TryGetValues("Set-Cookie", out var setCookies))
                context.Response.Headers.SetCookie = setCookies.Where(value => value.StartsWith(CookieName + "=", StringComparison.Ordinal)).ToArray();
            await upstream.Content.CopyToAsync(context.Response.Body, context.RequestAborted);
        }
        catch (HttpRequestException)
        {
            if (!context.Response.HasStarted) await RefuseAsync(context, 502, "selected_session_upstream_unavailable");
            else context.Abort();
        }
    }

    private static Task RefuseAsync(HttpContext context, int status, string error)
    {
        context.Response.StatusCode = status;
        context.Response.Headers.CacheControl = "no-store";
        return context.Response.WriteAsJsonAsync(new { error }, context.RequestAborted);
    }

    public void Dispose() => _sender.Dispose();
}
